### PR TITLE
Integration of fp16 implicit-gemm bwd dynamic asm kernel for gfx908

### PR DESCRIPTION
**Pull request by fronteer**
### Add dynamic asm igemm kernel for conv-bwd with fp16 data type on gfx908.

<details><summary><b><i>The script used to test implicit-gemm bwd-fp16 dynamic with MIOpen</i></b></summary>

```bash
#!/bin/bash

#flag enables fwd, bwd, wrw convolutions
FORW=2
Verify=1

set -x

#resnext101
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 1024 -H 14 -W 14 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 1024 -H 14 -W 14 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 1024 -H 14 -W 14 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 2048 -H 7 -W 7 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 256 -H 56 -W 56 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 256 -H 56 -W 56 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
##../Build/bin/MIOpenDriver convfp16 -n 64 -c 256 -H 56 -W 56 -k 512 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1

#test for group > 1
../Build/bin/MIOpenDriver convfp16 -n 64 -c 1024 -H 14 -W 14 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 2 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 64 -c 2048 -H 7 -W 7 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 2 -F $FORW -V $Verify -t 1

#efficiency compare with hip
../Build/bin/MIOpenDriver convfp16 -n 256 -c 1024 -H 14 -W 14 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 1024 -H 14 -W 14 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 1024 -H 14 -W 14 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 128 -H 28 -W 28 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 128 -H 28 -W 28 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 128 -H 58 -W 58 -k 128 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 2048 -H 7 -W 7 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 14 -W 14 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 14 -W 14 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 30 -W 30 -k 256 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 56 -W 56 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 56 -W 56 -k 512 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 256 -H 56 -W 56 -k 64 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 16 -W 16 -k 512 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 28 -W 28 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 28 -W 28 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 28 -W 28 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 7 -W 7 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 512 -H 7 -W 7 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 64 -H 56 -W 56 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 64 -H 56 -W 56 -k 64 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 64 -H 56 -W 56 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 256 -c 3 -H 230 -W 230 -k 64 -y 7 -x 7 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 1024 -H 14 -W 14 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 1024 -H 14 -W 14 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 1024 -H 14 -W 14 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 128 -H 28 -W 28 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 128 -H 28 -W 28 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 128 -H 58 -W 58 -k 128 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 2048 -H 7 -W 7 -k 512 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 14 -W 14 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 14 -W 14 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 30 -W 30 -k 256 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 56 -W 56 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 56 -W 56 -k 512 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 256 -H 56 -W 56 -k 64 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 16 -W 16 -k 512 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 28 -W 28 -k 1024 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 28 -W 28 -k 128 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 28 -W 28 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 7 -W 7 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 512 -H 7 -W 7 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 64 -H 56 -W 56 -k 256 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 64 -H 56 -W 56 -k 64 -y 1 -x 1 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 64 -H 56 -W 56 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1
../Build/bin/MIOpenDriver convfp16 -n 128 -c 3 -H 230 -W 230 -k 64 -y 7 -x 7 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F $FORW -V $Verify -t 1

set +x 

```
</details>

<details><summary><b><i>Environments needed for running the testing </i></b></summary>

```bash
#!/bin/bash
unset MIOPEN_DEBUG_CONV_WINOGRAD
unset MIOPEN_DEBUG_CONV_FFT
unset MIOPEN_DEBUG_CONV_DIRECT
unset MIOPEN_DEBUG_CONV_GEMM
unset MIOPEN_DEBUG_CONV_SCGEMM
unset MIOPEN_DEBUG_CONV_IMPLICIT_GEMM
unset MIOPEN_DEBUG_FIND_ONLY_SOLVER
unset MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_FIRST_SOLUTION
unset MIOPEN_FIND_ENFORCE
unset MIOPEN_LOG_LEVEL
unset MIOPEN_FIND_MODE
unset MIOPEN_DEBUG_FIND_ONLY_SOLVER

export MIOPEN_DEBUG_CONV_WINOGRAD=0
export MIOPEN_DEBUG_CONV_FFT=0
export MIOPEN_DEBUG_CONV_DIRECT=0
export MIOPEN_DEBUG_CONV_GEMM=0
export MIOPEN_DEBUG_CONV_SCGEMM=0
export MIOPEN_DEBUG_CONV_IMPLICIT_GEMM=1

##export MIOPEN_LOG_LEVEL=4

export MIOPEN_FIND_MODE=NORMAL
export MIOPEN_DEBUG_FIND_ONLY_SOLVER=ConvAsmImplicitGemmGTCDynamicBwdXdlops
```

</details>

<details><summary><b><i> Performance Data</i></b></summary>

[Asm bwd-fp16 Performance](https://amdcloud-my.sharepoint.com/:x:/r/personal/qfzhang_amd_com/Documents/MI100(asm-generator)-fp16-bwd-results.xlsx?d=w90149c64e05e4994ba6fe3279002f40c&csf=1&web=1&e=PkxT0X)

</details>

### CODE REVIEW

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp [hunk 1]
```diff
@@ -163,23 +163,707 @@ GetImplicitGemmGtcDynamicBwdTunablesList()
...
+static bool mayHaveBiggerGemmNPerBlock(int gemm_m,
+                                       int gemm_n,
+                                       bool is_unit_yx,
+                                       const TunableImplicitGemmGTCDynamic_t& cfg)
+{
+    float n_times_m = static_cast<float>(gemm_n) / static_cast<float>(gemm_m);
+
+    // since gemm_n is much bigger than gemm_m, we want to use a config where gemm_n_per_block is
+    // bigger than gemm_m_per_block
+    if(!is_unit_yx && n_times_m >= 16.0f && cfg.gemm_m_per_block >= cfg.gemm_n_per_block &&
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:844]: How to get the 16.f here?

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp [hunk 2]
```diff
@@ -163,23 +163,707 @@ GetImplicitGemmGtcDynamicBwdTunablesList()
...
                   std::string, // kernel_name
                   int,         // block_size
                   int>         // grid_size
     FindImplicitGemmGtcDynamicBwdKernel(const ConvolutionContext& ctx)
 {
-    auto tunables         = GetImplicitGemmGtcDynamicBwdTunablesList();
+    auto tunables = GetImplicitGemmGtcDynamicBwdTunablesList(ctx);
+
+    // so far, "group" is only supported by bwd fp16 kernels
+    const auto group      = ctx.IsFp16() ? ctx.group_counts : 1;
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:861]: How to disable the fp32 group convolution case?

File: src/kernels/dynamic_igemm/igemm_gtc_bwd_fp16/igemm_upsampling_clear_nchw_fp16.inc [hunk 3]
```diff
@@ -0,0 +1,195 @@
+/*******************************************************************************
```
Comment (shaojiewang) [inline src/kernels/dynamic_igemm/igemm_gtc_bwd_fp16/igemm_upsampling_clear_nchw_fp16.inc:1]: Please do not commit the asm file which is never used.

File: src/kernels/dynamic_igemm/igemm_gtc_bwd_fp16/igemm_bwd_gtc_gfx908_fp16_004x064.inc [hunk 4]
```diff
@@ -0,0 +1,1400 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
```
Comment (shaojiewang) [inline src/kernels/dynamic_igemm/igemm_gtc_bwd_fp16/igemm_bwd_gtc_gfx908_fp16_004x064.inc:5]: please update license

**Review by shaojiewang — CHANGES_REQUESTED** [review]

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 2 above)
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:861]: Right now all group convolutions disabled for this solver in IsApplicable:
```C++
    if(ctx.group_counts != 1)
        return false;
```

This MUST be updated to allow group convolutions for FP16.

**Review by atamazov — DISMISSED** [review]
Please describe how this solver was tested (add to PR description), pay attention to grouped FP16 cases.

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 2 above)
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:861]: This opens a question how group FP16 convolutions were tested by @fronteer

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 1 above)
Comment (fronteer) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:844]: Experimental. No theoretical method to fix a threshold for this.

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 2 above)
Comment (fronteer) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:861]: Fixed and tested with group for bwd-fp16

**Comment by atamazov** [comment]
@fronteer I highly recommend using valid github markup in the description.

**Comment by atamazov** [comment]
[Off-topic] @fronteer This is how I insert console logs:

### Short logs

Enclose logs into ``` -- that would ensure good visual emphasis and absence of word wrapping:
````
```bash
INSERT LOG CONTENTS HERE
```
````

### Long logs

````
<details><summary><b><i>HEADLINE</i></b></summary>

```bash
INSERT LOG CONTENTS HERE
```
</details>
````
Notice one empty line after `</summary>` -- this is necessary for valid formatting.

### Very long logs

Attach as a .txt file.

**Comment by atamazov** [comment]
[Off-topic] Created wiki page: https://github.com/ROCmSoftwarePlatform/MIOpen/wiki/How-to-insert-console-logs-into-github-pages

**Comment by asroy** [comment]
@fronteer Please add link to perf number

**Comment by atamazov** [comment]
Assigned reviewers, please go ahead.

**Review by shaojiewang — APPROVED** [review]

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp [hunk 5]
```diff
@@ -29,12 +29,75 @@
...
 #include <miopen/conv/context.hpp>
+#include <miopen/numeric.hpp>
+#include <miopen/conv/asm_implicit_gemm.hpp>
 
 #include <vector>
 
 namespace miopen {
 namespace conv {
 
+struct _magic_div_u32_
```
Comment (atamazov) [inline src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:40]: I've seen this in #680

**Review by atamazov — COMMENTED** [review]
#680 must go first. Then I expect merge conflicts.

**Comment by fronteer** [comment]
> #680 must go first. Then I expect merge conflicts.

Yes, we should have added magic division as a common utility.

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. Please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp (hunk 5 above)
Comment (asroy) [reply src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:40]: So this should be moved into a separate utility file and share it, right?

@atamazov @fronteer

**Comment by atamazov** [comment]
@fronteer #680 is merged in, so shall we continue with this PR?

**Comment by fronteer** [comment]
> @fronteer #680 is merged in, so shall we continue with this PR?

Yes.  I will update soon

File: src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp (hunk 5 above)
Comment (atamazov) [reply src/include/miopen/conv/invokers/impl_gemm_dynamic.hpp:40]: No. Just a warning. Now I see that #680 has is in the same file. Nothing to do.

**Comment by atamazov** [comment]
Is it blocker for 4.3 release?

**Comment by atamazov** [comment]
Unfortunately, this should wait until #858 is merged in. #680 is reverted in the Mainline.

**Comment by fronteer** [comment]
This has just passed the CI test.

**Comment by shaojiewang** [comment]
Can the draft flag be removed?

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 6]
```diff
@@ -215,6 +217,184 @@ InvokerFactory MakeImplGemmDynamicBackwardDataInvokerFactory(const ConvolutionCo
...
+                SetTensor(handle, tensors.outDesc, tensors.out, &zero);
+
+                if(handle.IsProfilingEnabled())
+                    elapsed += handle.GetKernelTime();
+            }
+            for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
+            {
+                if(is_gemm_not_empty[gemm_id])
+                {
+                    magic_div_u32_t mdiv_0 =
```
Comment (shaojiewang) [inline src/conv/invokers/impl_gemm_dynamic.cpp:338]: Can mdiv_0 be computed out of the invoker? `magic_div_u32_gen` computes a integer division.

File: src/conv/invokers/impl_gemm_dynamic.cpp [hunk 7]
```diff
@@ -215,6 +217,184 @@ InvokerFactory MakeImplGemmDynamicBackwardDataInvokerFactory(const ConvolutionCo
...
+                    uint32_t shift_pack_1 =
+                        magic_div_u32_pack_shift(mdiv_4.shift, mdiv_5.shift, mdiv_6.shift, 0);
+
+                    handle.Run(kernel)(tensors.out,
+                                       tensors.w,
+                                       tensors.in,
+                                       hi,
+                                       wi,
+                                       n,
+                                       k / group,
```
Comment (shaojiewang) [inline src/conv/invokers/impl_gemm_dynamic.cpp:357]: k/group can also be moved out of invoker.

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp [hunk 8]
```diff
@@ -260,66 +875,87 @@ static std::tuple<bool,        // is suitable kernel found
...
-    // second try, try find if packed image size match
-    for(const auto& cfg : tunables)
+    if(ctx.IsFp32())
     {
-        const auto b = cfg.nxe == 0
-                           ? h_tilda_slice * w_tilda_slice
-                           : ((h_tilda_slice * w_tilda_slice + cfg.nxb - 1) / cfg.nxb) * cfg.nxb;
-        const auto gemm_n_packed = n * b;
-        if(cfg.nxe == 0)
+        // second try, try find if packed image size match
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:908]: `gemm_n_packed` has been check before this piece of code, why it need a second try?

**Review by shaojiewang — COMMENTED** [review]
Just some little suggestions.

**Comment by shaojiewang** [comment]
Well, I turned this PR to OPEN.

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 8 above)
Comment (fronteer) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:908]: My commit does not  touch the stuff used by fp32,  so in the first try, gemm_n_packed is the same as gemm_n

**Comment by atamazov** [comment]
Unfortunately, the PR description does not tell us how it affects the performance of the library (as a whole). So I need I need to perform some testing, in addition to reviewing.

**Comment by atamazov** [comment]
Please resolve merge conflicts.

**Comment by fronteer** [comment]
commit [51b5](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/710/commits/2721c946fb02ccf6f51f13415b7d7cf1cbe051b5) fixes a bug that causes the failure of some bwd fp32 cases.

**Comment by fronteer** [comment]
commit [51b5](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/710/commits/2721c946fb02ccf6f51f13415b7d7cf1cbe051b5)

**Comment by fronteer** [comment]
Most recent commit [ec85](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/710/commits/ccd90077b6d82ac018529d937396275221bbec85) passed the CI

File: src/conv/invokers/impl_gemm_dynamic.cpp (hunk 6 above)
Comment (shaojiewang) [reply src/conv/invokers/impl_gemm_dynamic.cpp]: It has been moved.

**Comment by atamazov** [comment]
@fronteer Of course but I need to obtain [some performance data](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/710#issuecomment-825682491) in order to merge this.

**Comment by atamazov** [comment]
## Performance testing results of non-DEV Release build

Preconditions:
- MI100 node
- BUILD_DEV=off, release build
- `asm-bwd-fp16` ccd90077b vs. baseline (`develop` b09568120)
- MIOPEN_FIND_MODE=NORMAL
- MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS=1
- MIOPEN_ENABLE_LOGGING_ELAPSED_TIME=1
- 93 FP16 configs.

Results:
- Find API, 1st run (filling the binary kernel cache)
  - FP16 Bwd GPU time: +10% on average, max ~230%
  - All other performance metrics: on par
- Find API, 2nd run (the binary kernel cache is full)
  - FP16 Bwd GPU time: +10% on average, max ~240%
  - All other performance metrics: on par
- Immediate mode API
  - FP16 Bwd GPU time: +10% on average, max ~200%
  - FP16 Bwd Wall Time: +10% on average
  - FP16 Bwd Aux Wall time: ~200% on average
    - [Off-topic] @JehandadKhan The reason of such a high performance gain is time spent for perf-db reads in the baseline. The fastest solver in baseline is `ConvHipImplicitGemmBwdDataV4R1Xdlops`. It needs performance-config, and reading it from the system-perf-db takes ~190ms. In `asm-bwd-fp16`, the fastest solver is `ConvAsmImplicitGemmGTCDynamicBwdXdlops` in many cases; it does not need performance-config. As a result, Aux Wall time for the former case is around 200 ms, but for the latter case it is milliseconds (see attached logs).
  - All other performance metrics: on par

Raw console logs and CSVs: [asm-bwd-fp16_man.zip](https://github.com/ROCmSoftwarePlatform/MIOpen/files/6392102/asm-bwd-fp16_man.zip)

**Comment by atamazov** [comment]
## Performance testing results of DEV Release build

Preconditions:
- MI100 node
- BUILD_DEV=On, release build
- `asm-bwd-fp16` ccd90077b vs. baseline (`develop` b09568120)
- MIOPEN_FIND_MODE=NORMAL
- MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS=1
- 93 FP16 configs, 93 FP32 configs.

Results:
- FP32, Find API and Immediate Mode API:
  - All performance metrics on par.
- FP16
  - Find API
    - Bwd, GPU Time and Wall time: approx. +10% on average
  - Immediate Mode API
    - Bwd
      - GPU Time and Wall time: approx. +10% on average
      - Aux Wall time: approx. +330% (3.3X) on average, max 185X.
  - All other performance metrics: on par

Logs and CSVs: [asm-bwd-fp16_BIULD_DEV.zip](https://github.com/ROCmSoftwarePlatform/MIOpen/files/6392799/asm-bwd-fp16_BIULD_DEV.zip)

**Comment by atamazov** [comment]
Performance testing verdict: Ok :green_circle: 

@fronteer Please find reports in two previous comments. This is roughly what I would like to see in all PRs that may change (or should improve) the performance of the library. Let me know if you have any questions.

/cc @junliume

**Comment by atamazov** [comment]
@carlushuang @asroy Could you please review this?

**Comment by JehandadKhan** [comment]
> The reason of such a high performance gain is time spent for perf-db reads in the baseline. The fastest solver in baseline is ConvHipImplicitGemmBwdDataV4R1Xdlops. It needs performance-config, and reading it from the system-perf-db takes ~190ms. In asm-bwd-fp16, the fastest solver is ConvAsmImplicitGemmGTCDynamicBwdXdlops in many cases; it does not need performance-config. As a result, Aux Wall time for the former case is around 200 ms, but for the latter case it is milliseconds (see attached logs).

@atamazov I believe this is due to the measurement being done with the driver, since each invocation of the driver re-initializes the database, however, in a network once the library is loaded in memory the db is initialized and this is not an issue.

**Review by shaojiewang — APPROVED** [review]

**Comment by fronteer** [comment]
> Performance testing verdict: Ok green_circle
> 
> @fronteer Please find reports in two previous comments. This is roughly what I would like to see in all PRs that may change (or should improve) the performance of the library. Let me know if you have any questions.
> 
> /cc @junliume

So this is to find the total performance effect to implicit-gemm when more kernel implementations are added for selecting/evaluating (MIOPEN_DEBUG_IMPLICIT_GEMM_FIND_ALL_SOLUTIONS=1)  ?

**Comment by atamazov** [comment]
@fronteer Yes, the idea is to evaluate the total performance effects related to GPU performance and host-side overheads (Wall time and Aux Wall time). The criterion is: The candidate PR must show some clear performance gain and none (or, small enough) additional performance overhead. In other words, the new Solver should be clearly faster than all existing ones at acceptable cost.

/cc @junliume

**Pull request merged**